sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -178,44 +191,58 @@ private Worker(
     }
 
     private void addSpan(ReadableSpan span) {
-      if (!queue.offer(span)) {
+      if (queueSize.get() >= maxQueueSize) {
         droppedSpans.add(1);
+      } else {
+        queue.offer(span);
+        queueSize.incrementAndGet();
+        if (addedSpansCounter.incrementAndGet() % maxExportBatchSize == 0) {
```

Comment:
can't we use the queueSize here, rather than having to keep another counter in sync?

Reply:
queue.size() is linear in complexity, so it is really bad. This is the reason I was proposing to use MPSCQueue which is much cleaner

Reply:
no, I was saying your `queueSize` variable which you're using to track the size of the queue, and could have kept the value of on the line above.

Reply:
queueSize changes because of exporter reading from the queue, so we cannot use this to determine whether a signal is needed. What we really want is a signal for the exporter every  maxExportBatchSize so exporter sends the entries as soon as possible.

An example with maxExportBatchSize=100 to show why queueSize is not great.

- Let's say we have 90 entries in the queue added. queueSize is 90.
- Exporter in parallel reads all the 90 entries. Now queueSize is 90.
- Exporter goes to sleep.
- We have 10 more new entries in the queue. queueSize is 10.
- If we use queueSize here, then exporter would wait for 90 more entries which introduces delay.

A simple counter that tracks total added spans is good enough.

Reply:
We should trigger an export in only 3 cases.
1) The size of the queue has been reached (`queueSize` is tracking that here)
2) The timer has expired without an export
3) flush() is called.

I don't see how using some other number (the total number of things added to the queue throughout its entire history) meets any of these 3 criteria. That number has very little to do with whether any of those 3 criteria have been met.

Reply:
Agreed with the three criteria except criteria (1) is not based on `queueSize` but on `batchSize`. Batch size tracks the entries in the buffer ready for export.

Triggering an export in the exporter thread and signaling exporter thread about new entries in the queue are two separate concerns. The three criteria you mentioned are being met in the exporter thread (run() function)

1. We check if the batch size reached maxExportBatchSize.
  `batch.size() >= maxExportBatchSize`
2. The timer has expired is verified in the same line as well. 
  `System.nanoTime() >= nextExportTime`
    Note that exporter thread only sleep to a max of scheduleDelayNanos to satisfy this criteria.
3. forceFlush() is sending a signal to the exporter thread with a flag, so it does flush() irrespective of the above 2 conditions.

Now coming back to the queue writer here is how criteria (1) is satisfied. Every maxExportBatchSize adds to the queue, we send a signal to the exporter thread about new entries. Exporter thread when woke up has the guarantee that it will read up maxExportBatchSize entries and trigger the export. We cannot simply use `queueSize == maxExportBatchSize` here because of the above example I pointed out with maxExportBatchSize=100.

- Let's say we have 90 entries in the queue added. queueSize is 90.
- Exporter in parallel reads all the 90 entries. Now queueSize is 90.
- Exporter goes to sleep.
- We have 10 more new entries in the queue. queueSize is 10.

If we use queueSize here, then exporter would wait for 90 more entries which introduces delay. So using a counter and signaling every maxExportBatchSize ensures criteria (1) is met asap.

Reply:
@jkwatson This reminds me of an issue I filed a long long time ago

https://github.com/open-telemetry/opentelemetry-specification/issues/849

Never got clarification on whether we're supposed to eagerly export when we have maxExportBatchSize items. But it's true that what are current code is doing so we should preserve it for now which I think is the intent of this line.